src/hotspot/share/opto/mulnode.cpp

```diff
@@ -1683,3 +1713,64 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
     return TypeLong::LONG;
   }
 }
+
+// Helper method to transform:
+// patterns similar to (v << 2) & 3 to 0
+// and
+// patterns similar to (v1 + (v2 << 2)) & 3 transformed to v1 & 3
+bool MulNode::AndIL_shift_and_mask(PhaseGVN* phase, Node* mask, Node* shift, BasicType bt) {
```

Comment:
1. The operand order is distracting and perhaps confusing, since mask is in(2) and shift is in(1).
2. The mask is not necessarily in(2), because it doesn't have to be a constant; it can be a bounded value.
3. I wrote an expanded comment.
4. There's a redundant NULL check.
5. You should use a different name for the adjusted `bt` value.

Here are the changes I suggest to this function:

```
// Given an expression (AndX shift mask) or (AndX mask shift),
// determine if the AndX must always produce zero, because the
// the shift (x<<N) is bitwise disjoint from the mask #M.
// The X in AndX must be I or L, depending on bt.
// Specifically, the following cases fold to zero,
// when the shift value N is large enough to zero out
// all the set positions of the and-mask M.
//   (AndI (LShiftI _ #N) #M) => #0
//   (AndL (LShiftL _ #N) #M) => #0
//   (AndL (ConvI2L (LShiftI _ #N)) #M) => #0
// The M and N values must satisfy ((-1 << N) & M) == 0.
// Because the optimization might work for a non-constant
// mask M, we check the AndX for both operand orders.
bool MulNode::AndIL_shift_and_mask(PhaseGVN* phase, Node* shift, Node* mask, BasicType bt, bool check_reverse) const {
  if (mask == NULL || shift == NULL) {
    return false;
  }
  const TypeInteger* mask_t = phase->type(mask)->isa_integer(bt);
  const TypeInteger* shift_t = phase->type(shift)->isa_integer(bt);
  if (mask_t == NULL || shift_t == NULL) {
    return false;
  }
+ BasicType shift_bt = bt;
+ if (bt == T_LONG && shift->Opcode() == Op_ConvI2L) {
    Node* val = shift->in(1);
    if (val == NULL) {
      return false;
    }
+   if (val->Opcode() == Op_LShiftI) {
+     shift_bt = T_INT;
+     shift = val;
+   }
  }
#s/bt/shift_bt/
  if (shift->Opcode() != Op_LShift(shift_bt)) {
+   if (check_reverse &&
+       (mask->Opcode() == Op_LShift(bt) ||
+        (bt == T_LONG && mask->Opcode() == Op_ConvI2L))) {
+     // try it the other way around
+     return AndIL_shift_and_mask(phase, shift, mask, bt, false);
+   }
    return false;
  }
  Node* shift2 = shift->in(2);
  if (shift2 == NULL) {
    return false;
  }
  const Type* shift2_t = phase->type(shift2);
  if (!shift2_t->isa_int() || !shift2_t->is_int()->is_con()) {
    return false;
  }

#s/bt/shift_bt/
  jint shift_con = shift2_t->is_int()->get_con() & ((shift_bt == T_INT ? BitsPerJavaInteger : BitsPerJavaLong) - 1);
  if ((1L << shift_con) > mask_t->hi_as_long() && mask_t->lo_as_long() >= 0) {
    return true;
  }

  return false;
}
```
Suggest signature change: